cloud/aws/services/ec2/vpc.go

```diff
@@ -17,41 +17,85 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ec2"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
 )
 
 type vpcs interface {
-	DescribeVpcs(*ec2.DescribeVpcsInput) (*ec2.DescribeVpcsOutput, error)
+	DescribeVpcs(input *ec2.DescribeVpcsInput) (*ec2.DescribeVpcsOutput, error)
+	CreateVpc(input *ec2.CreateVpcInput) (*ec2.CreateVpcOutput, error)
+	DeleteVpc(input *ec2.DeleteVpcInput) (*ec2.DeleteVpcOutput, error)
 }
 
 type VPC struct {
-	ID string
+	ID        string
+	CidrBlock string // TODO(vincepri): use *net.IPNet?
 }
 
-func (s *Service) ReconcileVPC(id string) (*VPC, error) {
+func (s *Service) ReconcileVPC(v *v1alpha1.VPC) (*VPC, error) {
 	// Does it exist and look in good working order? ok exit no error
-	vpc, err := s.lookupVPCByID(id)
+	vpc, err := s.lookupVPCByID(v.ID)
 	if err != nil {
+		if err == ErrVPCNotFound {
+			return s.createVPC(v)
+		}
+
 		return nil, err
 	}
-	// If it doesn't exist, create it
+
+	if v.CidrBlock != vpc.CidrBlock {
+		// TODO(vincepri): delete and recreate?
+	}
+
 	return vpc, nil
 }
 
+func (s *Service) createVPC(v *v1alpha1.VPC) (*VPC, error) {
+	input := &ec2.CreateVpcInput{
+		CidrBlock: aws.String(v.CidrBlock),
+	}
+
+	out, err := s.VPCs.CreateVpc(input)
+	if err != nil {
+		return nil, err
+	}
+
+	return &VPC{
+		ID:        *out.Vpc.VpcId,
+		CidrBlock: *out.Vpc.CidrBlock,
+	}, nil
+}
+
+func (s *Service) deleteVPC(v *v1alpha1.VPC) error {
+	input := &ec2.DeleteVpcInput{
+		VpcId: aws.String(v.ID),
+	}
+
+	_, err := s.VPCs.DeleteVpc(input)
+	return err
+}
+
 func (s *Service) lookupVPCByID(id string) (*VPC, error) {
 	input := &ec2.DescribeVpcsInput{
 		VpcIds: []*string{
 			&id,
 		},
 	}
+
 	out, err := s.VPCs.DescribeVpcs(input)
```

Comment:
We should handle the real ec2 aws errors here and either return them or handle them. That way we won't need to define custom errors (probably)

Reply:
I think returning the errors might be ok here? From the [AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVpcs.html) these calls can return generic client errors, which may require user action. Alerting the user might be the first approach, we can always revisit it later.